docs/developer/api-versioning.md

```diff
@@ -0,0 +1,391 @@
+        :> Summary "Get a conversation by ID"
+        :> ZLocalUser
+        :> "conversations"
+        :> QualifiedCapture "cnv" ConvId
+        :> Get '[Servant.JSON] Conversation,
+    [...]
+  }
+```
+
+APIs of all the supported versions can be composed like this:
+
+```haskell
+type ServantAPI =
+  ToServantApi (Api "v1")
+    :<|> ToServantApi (Api "v2")
+    :<|> ToServantApi (Api "v4") -- v3 is broken
+```
+
+
+## Changes between versions
+
+The point of having versions is of course not that all of them look
+exactly the same except for their prefix.  The point is that there are
+other things that change between versions.
+
+There are essentially two categories of changes:
+
+1. **data**: request or response bodies, variable path segments,
+   possible headers or status codes (as in `UVerb` or `MultiVerb`),
+   etc.
+2. **structure**: literal path segments, verb, the version itself, ...
+
+
+## Changes in the data
+
+If a data type in request, response, variable path segments, or anywhere else
+changes, introduce a type family parameterized in the version.
+
+```haskell
+[...]
+    getConversation ::
+      routes
+        :- version
+        :> Summary "Get a conversation by ID"
+        :> ZLocalUser
+        :> "conversations"
+        :> QualifiedCapture "cnv" ConvId
+        :> Get '[Servant.JSON] (ConversationV version),
+[...]
+
+type family ConversationV (version :: Symbol) :: * where
+  ConversationV "v1" = Conversation
+  ConversationV "v2" = Conversation
+  ConversationV "v4" = ConversationV4
+```
+
+Note that before version `"v4"`, nothing changed for this type, so
+there was no need to introduce a new concrete data type.
+
+If the last change of a data type is entirely phased out, the type
+family turns constant and can be removed again.  If you see this in
+your code:
+
+```haskell
+type family ConversationV (version :: Symbol) :: * where
+  ConversationV "v4" = ConversationV4
+  ConversationV "v5" = ConversationV4
+  ConversationV "v6" = ConversationV4
+  ConversationV "v7" = ConversationV4
+```
+
+You can remove 'ConversationV', rename `ConversationV4` to Conversation,
+and use it in the routing table instead of `ConversationV` again, as
+before `"v4"`.
+
+
+## Changing structure
+
+Without loss of generality, we only consider additions and deletions
+of routes in this section: if you want to change the path or verb of
+an end-point, add a new path instead, and phase the old one out (now
+or in some future version).
+
+When end-points are present in some supported versions, but not in
+others, their record fields in the servant routing type needs to be
+present for all versions, but in some versions should behave as if it
+weren't.
+
+This is best solved by a new type alias:
+
+```haskell
+type NotInThisVersion = Verb 'NOTINTHISVERSION '[] NoContent
+```
+
+Then we can write a type family that can crawl a `ServantAPI` (not the
+record one, the one with `:<|>`) and drop all the routes marked as not
+existing.
```

Comment:
if end-points do not change behavior accross all supported versions, there is no need for type families.  if the last changed version is phased out, the resp. type family can be removed.

the only places in the routing table that use the version argument in any other way than path prefix are those that have different behavior on different supported versions.  it it's just a json body that changes, than the type of that body becomes a type family.  if it's a route that has been added or removed, the versions not supporting it get `NotInThisVersion` as a verb.

Reply:
I don't consider this thought essential to the approach, but I think I would like to fight for it a little more since I find it pretty elegant and clean.  Let's maybe have a call?